File: app/src/main/java/com/cookplan/add_ingredient_view/ProductListAdapter.java
package com.cookplan.add_ingredient_view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.TextView;

import com.cookplan.R;
import com.cookplan.RApplication;
import com.cookplan.models.Product;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by DariaEfimova on 20.03.17.
 */

public class ProductListAdapter extends ArrayAdapter<Product> {
    private List<Product> items;
    private List<Product> itemsAll;
    private List<Product> suggestions;

    public ProductListAdapter(Context context, List<Product> items) {
        super(context, R.layout.product_autocomplete_item_layout, items);
        this.items = items;
        this.itemsAll = new ArrayList<>(items);
        this.suggestions = new ArrayList<>();
    }

    @NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        View v = convertView;
        if (v == null) {
            LayoutInflater vi = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            v = vi.inflate(R.layout.product_autocomplete_item_layout, null);
        }
        Product product = items.get(position);
        if (product != null) {
            TextView productNameTextView = (TextView) v.findViewById(R.id.item_name);
            if (productNameTextView != null) {
//              Log.i(MY_DEBUG_TAG, "getView Customer Name:"+customer.getName());
                productNameTextView.setText(product.getName());
            }
            if (product.getCategory() != null) {
                TextView categoryNameTextView = (TextView) v.findViewById(R.id.category_item_name);
                if (categoryNameTextView != null) {
                    categoryNameTextView.setText(product.getCategory().toString());
                    categoryNameTextView.setTextColor(ContextCompat.getColor(RApplication.getAppContext(),
                            product.getCategory().getColorId()));
                }
                View categoryView = v.findViewById(R.id.category_view);
                if (categoryView != null) {
                    categoryView.setBackgroundColor(ContextCompat.getColor(RApplication.getAppContext(),
                            product.getCategory().getColorId()));
                }
            }

        }
        return v;
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return nameFilter;
    }

    private Filter nameFilter = new Filter() {
        @Override
        public String convertResultToString(Object resultValue) {
            String str = ((Product) (resultValue)).getName();
            return str;
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            if (constraint != null) {
                suggestions.clear();
                for (Product product : itemsAll) {
                    if (product.getName().toLowerCase().startsWith(constraint.toString().toLowerCase())) {
                        suggestions.add(product);
                    }
                }
                FilterResults filterResults = new FilterResults();
                filterResults.values = suggestions;
                filterResults.count = suggestions.size();
                return filterResults;
            } else {
                return new FilterResults();
            }
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            ArrayList<Product> filteredList = (ArrayList<Product>) results.values;
            if (results.count > 0) {
                clear();
                for (Product c : filteredList) {
                    add(c);
                }
                notifyDataSetChanged();
            }
        }
    };

}

File: app/src/main/java/com/cookplan/recipe/import_recipe/approve_result/ProductListAdapter.kt
package com.cookplan.recipe.import_recipe.approve_result

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.cookplan.R
import com.cookplan.models.Product
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Created by DariaEfimova on 20.03.17.
 */

class ProductListAdapter(context: Context, private val itemsAll: List<Product>) : ArrayAdapter<Product>(context, R.layout.product_autocomplete_item_layout, itemsAll) {
    private val suggestions: MutableList<Product> = mutableListOf()

    private val nameFilter = object : Filter() {
        override fun convertResultToString(resultValue: Any): String {
            return (resultValue as Product).toStringName()
        }

        override fun performFiltering(constraint: CharSequence?): Filter.FilterResults {
            suggestions.clear()
            for (product in itemsAll) {
                if (product.toStringName().toLowerCase().contains(constraint.toString().toLowerCase())) {
                    suggestions.add(product)
                }
            }
            val filterResults = Filter.FilterResults()
            filterResults.values = suggestions
            filterResults.count = suggestions.size
            return filterResults
        }

        override fun publishResults(constraint: CharSequence?, results: Filter.FilterResults) {
            if (results.values != null && results.count > 0) {
                val filteredList = CopyOnWriteArrayList(results.values as ArrayList<*>)
                clear()
                for (c in filteredList) {
                    add(c as Product)
                }
                notifyDataSetChanged()
            }
        }
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val vi = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        val view = convertView ?: vi.inflate(R.layout.product_autocomplete_item_layout, null)

        val product = itemsAll[position]

        val productNameTextView = view?.findViewById<TextView>(R.id.product_item_name)
        productNameTextView?.text = product.toStringName()

        val categoryNameTextView = view.findViewById<TextView>(R.id.category_product_item_name)
        categoryNameTextView?.setText(product.category.getNameResourceId())
        categoryNameTextView?.setTextColor(ContextCompat.getColor(parent.context, product.category.colorId))

        val categoryView = view.findViewById<View>(R.id.category_view)
        categoryView?.setBackgroundResource(product.category.colorId)

        return view
    }

    override fun getFilter(): Filter {
        return nameFilter
    }
}